modules/version-checker/version_checker.go

```diff
@@ -0,0 +1,174 @@
+	"regexp"
+)
+
+// VersionMismatchErr is an error to indicate version mismatch.
+type VersionMismatchErr struct {
+	expectedVersion string
+	actualVersion   string
+}
+
+func (r *VersionMismatchErr) Error() string {
+	return fmt.Sprintf("found version mismatch {%s} when "+
+		"expecting min version of {%s}", r.expectedVersion, r.actualVersion)
+}
+
+// VersionCheckerBinary is an enum for supported version checking.
+type VersionCheckerBinary int
+
+// List of binaries supported for version checking.
+const (
+	Docker VersionCheckerBinary = iota
+	Terraform
+	Packer
+)
+
+const (
+	// versionRegexMatcher is a regex used to extract version string from shell command output.
+	versionRegexMatcher = `\d+(\.\d+)+`
+	// defaultVersionArg is a default arg to pass in to get version output from shell command.
+	defaultVersionArg = "--version"
+)
+
+type CheckVersionParams struct {
+	Binary          VersionCheckerBinary
+	ExpectedVersion string
+	WorkingDir      string
+}
+
+// CheckVersionE checks whether the given Binary version is greater than or equal
+// to the given expected version.
+func CheckVersionE(
+	t testing.TestingT,
+	params CheckVersionParams) error {
+
+	if err := validateParams(params); err != nil {
+		return err
+	}
+
+	binaryVersion, err := getVersionWithShellCommand(t, params)
+	if err != nil {
+		return err
+	}
+
+	if err := checkMinimumBinaryVersion(binaryVersion, params.ExpectedVersion); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CheckVersion checks whether the given Binary version is greater than or equal to the
+// given expected version and fail if it's not.
+func CheckVersion(
+	t testing.TestingT,
+	params CheckVersionParams) {
+	require.NoError(t, CheckVersionE(t, params))
+}
+
+// Validate whether the given params contains valid data to check version.
+func validateParams(params CheckVersionParams) error {
+	// Check for empty parameters
+	if params.ExpectedVersion == "" {
+		return fmt.Errorf("set ExpectedVersion in params")
+	} else if params.WorkingDir == "" {
+		return fmt.Errorf("set WorkingDir in params")
+	} else if params.Binary < 0 {
+		return fmt.Errorf("set Binary in params")
+	}
+
+	// Check the format of the expected version.
+	if _, err := version.NewVersion(params.ExpectedVersion); err != nil {
+		return fmt.Errorf(
+			"invalid version format found {%s}", params.ExpectedVersion)
+	}
+
+	return nil
+}
+
+// getVersionWithShellCommand get version by running a shell command.
+func getVersionWithShellCommand(t testing.TestingT, params CheckVersionParams) (string, error) {
+	// Set appropriate Binary, versionArg and extract version function
+	// based on the given parameters.
+	var binaryName = ""
+	var versionArg = defaultVersionArg
+	switch params.Binary {
+	case Docker:
+		binaryName = "docker"
+	case Packer:
+		binaryName = "packer"
+	case Terraform:
+		binaryName = "terraform"
+	default:
+		t.Fatalf("unsupported Binary for checking versions {%s}.", params.Binary)
+	}
```

Comment:
Ah I just realized that using the `enum` here to figure out the binary to use may not be the best approach. This is specifically true for `terraform`, where the `Options` struct supports arbitrary terraform binary paths (https://github.com/gruntwork-io/terratest/blob/master/modules/terraform/options.go#L41)

The main use case for this is multi version projects, where users have different ways of installing multiple terraform versions and explicitly will point the binary path to something else.

Perhaps the struct should have two fields, one that is `BinaryType` which is the enum, and a `BinaryPath` field that is optional? And use this logic only if `BinaryPath` is empty?

Reply:
SGTM. Updated.